base/utfconvert.jl

```diff
@@ -0,0 +1,546 @@
+        buf[out += 1] = ch
+    end
+    UTF16String(buf)
+end
+
+convert(::Type{UTF8String},  dat::Vector{Char})   = convert(UTF8String, reinterpret(UInt32, dat))
+
+utf16(x) = convert(UTF16String, x)
+
+function convert(::Type{UTF16String}, str::ASCIIString)
+    dat = str.data
+    fast_utf_copy(UInt16, length(dat)+1, dat)
+end
+
+function convert(::Type{UTF32String}, str::ASCIIString)
+    dat = str.data
+    fast_utf_copy(Char, length(dat)+1, dat)
+end
+
+convert(::Type{UTF16String}, str::UTF16String)    = str
+convert(::Type{UTF16String}, dat::Vector{Char})   = convert(UTF16String, reinterpret(UInt32, dat))
+
+convert(::Type{Vector{UInt16}}, str::UTF16String) = str.data
+convert(::Type{Array{UInt16}},  str::UTF16String) = str.data
+
+utf32(x) = convert(UTF32String, x)
+
+convert(::Type{UTF32String}, str::UTF32String)    = str
+
+convert(::Type{UTF32String}, c::Char)             = UTF32String(Char[c, Char(0)])
+
+unsafe_convert{T<:Union(Int16,UInt16)}(::Type{Ptr{T}}, s::UTF16String) =
+    convert(Ptr{T}, pointer(s))
+
+function convert(::Type{UTF16String}, data::AbstractVector{UInt16})
+    !isvalid(UTF16String, data) && throw(ArgumentError("invalid UTF16 data"))
+    len = length(data)
+    @inbounds return UTF16String(setindex!(copy!(Vector{UInt16}(len+1),1,data,1,len),0,len+1))
+end
+
+convert(T::Type{UTF16String}, data::AbstractArray{UInt16}) =
+    convert(T, reshape(data, length(data)))
+
+convert(T::Type{UTF16String}, data::AbstractArray{Int16}) =
+    convert(T, reinterpret(UInt16, data))
+
+function convert(T::Type{UTF16String}, bytes::AbstractArray{UInt8})
+    isempty(bytes) && return UTF16String(UInt16[0])
+    isodd(length(bytes)) && throw(ArgumentError("odd number of bytes"))
+    data = reinterpret(UInt16, bytes)
+    # check for byte-order mark (BOM):
+    if data[1] == 0xfeff        # native byte order
+        d = Array(UInt16, length(data))
+        copy!(d,1, data,2, length(data)-1)
+    elseif data[1] == 0xfffe    # byte-swapped
+        d = Array(UInt16, length(data))
+        for i = 2:length(data)
+            d[i-1] = bswap(data[i])
+        end
+    else
+        d = Array(UInt16, length(data) + 1)
+        copy!(d,1, data,1, length(data)) # assume native byte order
+    end
+    d[end] = 0 # NULL terminate
+    !isvalid(UTF16String, d) && throw(ArgumentError("invalid UTF16 data"))
+    UTF16String(d)
+end
+
+utf16(p::Ptr{UInt16}, len::Integer) = utf16(pointer_to_array(p, len))
+utf16(p::Ptr{Int16}, len::Integer) = utf16(convert(Ptr{UInt16}, p), len)
+function utf16(p::Union(Ptr{UInt16}, Ptr{Int16}))
+    len = 0
+    while unsafe_load(p, len+1) != 0; len += 1; end
+    utf16(p, len)
+end
+
+function convert(::Type{UTF32String}, data::AbstractVector{Char})
+    len = length(data)
+    @inbounds return UTF32String(setindex!(copy!(Vector{Char}(len+1),1,data,1,len),0,len+1))
+end
+
+convert{T<:Union(Int32,UInt32)}(::Type{UTF32String}, data::AbstractVector{T}) =
+    convert(UTF32String, reinterpret(Char, data))
```

Comment:
I'm not sure reinterpret makes sense to call on an AbstractVector. It might be better to combine with the above method signature to make on method definition.